### PR TITLE
Implement Version-Control-Kit (again)

### PR DESCRIPTION
**Pull request by 0xWDG**
## Summary

Re-implement version control kit.
Main source: https://github.com/AuroraEditor/AuroraEditor/commit/eb46902b49ce773d04c020a998cd541f55154989

Depended on a fix for https://github.com/AuroraEditor/Version-Control-Kit/issues/22
which is implemented now

## Changes Made

See: https://github.com/AuroraEditor/AuroraEditor/commit/eb46902b49ce773d04c020a998cd541f55154989

## TODO

None

## Motivation

https://github.com/AuroraEditor/AuroraEditor/commit/eb46902b49ce773d04c020a998cd541f55154989

## Checklist

Please ensure all of the following are completed before submitting the PR:

- [x] Code has been tested and verified.
- [x] Documentation has been updated to reflect the changes.
- [x] All tests pass successfully.
- [x] Code follows the project's coding guidelines and style.
- [x] The branch is up-to-date with the latest changes from the main branch.
- [x] Reviewed by at least one other contributor.

## Related Issues

This PR addresses the following issue(s): N/A

## Additional Notes

N/A

### CODE REVIEW

**Comment by aurora-care-bear** [comment]
Thanks for submitting a pull request!
We will review it as soon as possible.

**Comment by 0xWDG** [comment]
iCloud enabled dir error

```
working directory doesn't exist.
(
	0   CoreFoundation                      0x000000018dd18540 __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x000000018d809eb4 objc_exception_throw + 60
	2   Foundation                          0x000000018ee1b258 -[NSConcreteTask launchWithDictionary:error:] + 4208
	3   Aurora Editor                       0x0000000100a73490 $s15Version_Control8GitShellV3git4args4path4name7optionsAA10IGitResultVSaySSG_10Foundation3URLVSSAA0J16ExecutionOptionsVSgtKF + 1828
	4   Aurora Editor                       0x0000000100a57b94 $s15Version_Control8RevParseV17getRepositoryType12directoryURLAA0fG0O10Foundation0I0V_tKF + 1020
	5   Aurora Editor                       0x0000000100a19c18 $s15Version_Control5CheckV20checkIfProjectIsRepo12workspaceURLSb10Foundation0J0V_tF + 80
	6   Aurora Editor                       0x00000001006f8788 $s12AuroraEditor18SourceControlModelC12workspaceURLAC10Foundation0G0V_tcfc + 1008
	7   Aurora Editor                       0x00000001006f8388 $s12AuroraEditor18SourceControlModelC12workspaceURLAC10Foundation0G0V_tcfC + 56
	8   Aurora Editor                       0x00000001009dc76c $s12AuroraEditor17WorkspaceDocumentC04initC5State33_065D1843F6EB21E67F597C06B178642ELLyy10Foundation3URLVKF + 376
	9   Aurora Editor                       0x00000001009ddf50 $s12AuroraEditor17WorkspaceDocumentC4read4from6ofTypey10Foundation3URLV_SStKF + 380
	10  Aurora Editor                       0x00000001009dee88 $s12AuroraEditor17WorkspaceDocumentC4read4from6ofTypey10Foundation3URLV_SStKFTo + 192
	11  AppKit                              0x00000001916e0bc8 -[NSDocument initWithContentsOfURL:ofType:error:] + 352
	12  AppKit                              0x00000001917324c0 -[NSDocumentController makeDocumentWithContentsOfURL:ofType:error:] + 204
	13  AppKit                              0x0000000191a090ec __97-[NSDocumentController makeDocumentWithContentsOfURL:alternateContents:ofType:completionHandler:]_block_invoke + 96
	14  AppKit                              0x0000000191a08fd4 -[NSDocumentController makeDocumentWithContentsOfURL:alternateContents:ofType:completionHandler:] + 300
	15  AppKit                              0x0000000191a084c4 __102-[NSDocumentController _openDocumentWithContentsOfURL:requestedBySourceApp:display:completionHandler:]_block_invoke + 1140
	16  AppKit                              0x0000000191a07188 __144-[NSDocumentController _coordinateReadingAndGetAlternateContentsForOpeningDocumentAtURL:resolvingSymlinks:thenContinueOnMainThreadWithAccessor:]_block_invoke_4 + 164
	17  AppKit                              0x0000000191a075d0 __144-[NSDocumentController _coordinateReadingAndGetAlternateContentsForOpeningDocumentAtURL:resolvingSymlinks:thenContinueOnMainThreadWithAccessor:]_block_invoke.371 + 304
	18  AppKit                              0x0000000191a073e8 __144-[NSDocumentController _coordinateReadingAndGetAlternateContentsForOpeningDocumentAtURL:resolvingSymlinks:thenContinueOnMainThreadWithAccessor:]_block_invoke.367 + 368
	19  AppKit                              0x0000000191a167b0 ___NSMainRunLoopPerformBlockInModes_block_invoke + 44
	20  CoreFoundation                      0x000000018dca3574 __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 28
	21  CoreFoundation                      0x000000018dca3488 __CFRunLoopDoBlocks + 356
	22  CoreFoundation                      0x000000018dca291c __CFRunLoopRun + 2440
	23  CoreFoundation                      0x000000018dca193c CFRunLoopRunSpecific + 608
	24  HIToolbox                           0x000000019826a448 RunCurrentEventLoopInMode + 292
	25  HIToolbox                           0x000000019826a284 ReceiveNextEventCommon + 648
	26  HIToolbox                           0x0000000198269fdc _BlockUntilNextEventMatchingListInModeWithFilter + 76
	27  AppKit                              0x0000000191480ed0 _DPSNextEvent + 660
	28  AppKit                              0x0000000191c6beec -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 716
	29  AppKit                              0x000000019147437c -[NSApplication run] + 476
	30  AppKit                              0x000000019144b640 NSApplicationMain + 880
	31  Aurora Editor                       0x00000001000a0b80 main + 12
	32  dyld                                0x000000018d8450e0 start + 2360
)

**Comment by 0xWDG** [comment]
FOUND THE ERROR, now the fix remains.

It is caused by

https://github.com/AuroraEditor/Version-Control-Kit/blob/main/Sources/Version-Control/Base/Core/GitShell.swift#L34-L161

```swift
    public func git( // swiftlint:disable:this function_body_length cyclomatic_complexity
        args: [String],
        path: URL,
        name: String,
        options: IGitExecutionOptions? = nil
    ) throws -> IGitResult {
```

Temporary fix is/was
add 
```swift
        throw NSError(domain: "disabled", code: 0)
```
to prevent execution

File: AuroraEditor/Base/Version Control/Model/Client/Core/GitClient.swift [hunk 1]
```diff
@@ -51,31 +68,89 @@ public class GitClient: ObservableObject { // swiftlint:disable:this type_body_l
     @Published var publishedAllBranchNames: [String] = []
 
     public func getCurrentBranchName() throws -> String {
-        let output = try shellClient.run(
-            "cd \(directoryURL.relativePath.escapedWhiteSpaces());" +
-            "git rev-parse --abbrev-ref HEAD"
-        )
-            .replacingOccurrences(of: "\n", with: "")
-        if output.contains("fatal: not a git repository") {
+        let result = try GitShell().git(args: ["rev-parse", "--abbrev-ref", "HEAD"],
```
Comment (nanashili) [inline AuroraEditor/Base/Version Control/Model/Client/Core/GitClient.swift:71]: There is a function to get the current branch name

**Review by nanashili — APPROVED** [review]

**Comment by nanashili** [comment]
> FOUND THE ERROR, now the fix remains.
> 
> 
> 
> It is caused by
> 
> 
> 
> https://github.com/AuroraEditor/Version-Control-Kit/blob/main/Sources/Version-Control/Base/Core/GitShell.swift#L34-L161
> 
> 
> 
> ```swift
> 
>     public func git( // swiftlint:disable:this function_body_length cyclomatic_complexity
> 
>         args: [String],
> 
>         path: URL,
> 
>         name: String,
> 
>         options: IGitExecutionOptions? = nil
> 
>     ) throws -> IGitResult {
> 
> ```
> 
> 
> 
> Temporary fix is/was
> 
> add 
> 
> ```swift
> 
>         throw NSError(domain: "disabled", code: 0)
> 
> ```
> 
> to prevent execution

Getting the project repository type is causing the issues.

**Comment by aurora-care-bear** [comment]
Hey @0xWDG,

Thanks for your contribution.
Your PR has been merged!
Your name will appear on https://auroraeditor.com/#contributors in a few moments, if you have contributed to the editor.

Thanks again for your contribution!

Kind Regards,
Aurora Editor Team.

**Pull request merged**